File: zh-Hant/codes/java/chapter_greedy/coin_change_greedy.java
/**
 * File: coin_change_greedy.java
 * Created Time: 2023-07-20
 * Author: krahets (krahets@163.com)
 */

package chapter_greedy;

import java.util.Arrays;

public class coin_change_greedy {
    /* 零錢兌換：貪婪 */
    static int coinChangeGreedy(int[] coins, int amt) {
        // 假設 coins 串列有序
        int i = coins.length - 1;
        int count = 0;
        // 迴圈進行貪婪選擇，直到無剩餘金額
        while (amt > 0) {
            // 找到小於且最接近剩餘金額的硬幣
            while (i > 0 && coins[i] > amt) {
                i--;
            }
            // 選擇 coins[i]
            amt -= coins[i];
            count++;
        }
        // 若未找到可行方案，則返回 -1
        return amt == 0 ? count : -1;
    }

    public static void main(String[] args) {
        // 貪婪：能夠保證找到全域性最優解
        int[] coins = { 1, 5, 10, 20, 50, 100 };
        int amt = 186;
        int res = coinChangeGreedy(coins, amt);
        System.out.println("\ncoins = " + Arrays.toString(coins) + ", amt = " + amt);
        System.out.println("湊到 " + amt + " 所需的最少硬幣數量為 " + res);

        // 貪婪：無法保證找到全域性最優解
        coins = new int[] { 1, 20, 50 };
        amt = 60;
        res = coinChangeGreedy(coins, amt);
        System.out.println("\ncoins = " + Arrays.toString(coins) + ", amt = " + amt);
        System.out.println("湊到 " + amt + " 所需的最少硬幣數量為 " + res);
        System.out.println("實際上需要的最少數量為 3 ，即 20 + 20 + 20");

        // 貪婪：無法保證找到全域性最優解
        coins = new int[] { 1, 49, 50 };
        amt = 98;
        res = coinChangeGreedy(coins, amt);
        System.out.println("\ncoins = " + Arrays.toString(coins) + ", amt = " + amt);
        System.out.println("湊到 " + amt + " 所需的最少硬幣數量為 " + res);
        System.out.println("實際上需要的最少數量為 2 ，即 49 + 49");
    }
}

File: en/codes/csharp/chapter_greedy/coin_change_greedy.cs
/**
* File: coin_change_greedy.cs
* Created Time: 2023-07-21
* Author: hpstory (hpstory1024@163.com)
*/

namespace hello_algo.chapter_greedy;

public class coin_change_greedy {
    /* Coin change: Greedy */
    int CoinChangeGreedy(int[] coins, int amt) {
        // Assume coins list is ordered
        int i = coins.Length - 1;
        int count = 0;
        // Loop for greedy selection until no remaining amount
        while (amt > 0) {
            // Find the smallest coin close to and less than the remaining amount
            while (i > 0 && coins[i] > amt) {
                i--;
            }
            // Choose coins[i]
            amt -= coins[i];
            count++;
        }
        // If no feasible solution is found, return -1
        return amt == 0 ? count : -1;
    }

    [Test]
    public void Test() {
        // Greedy: can ensure finding a global optimal solution
        int[] coins = [1, 5, 10, 20, 50, 100];
        int amt = 186;
        int res = CoinChangeGreedy(coins, amt);
        Console.WriteLine("\ncoins = " + coins.PrintList() + ", amt = " + amt);
        Console.WriteLine("The minimum number of coins required to make up " + amt + " is" + res);

        // Greedy: cannot ensure finding a global optimal solution
        coins = [1, 20, 50];
        amt = 60;
        res = CoinChangeGreedy(coins, amt);
        Console.WriteLine("\ncoins = " + coins.PrintList() + ", amt = " + amt);
        Console.WriteLine("The minimum number of coins required to make up " + amt + " is" + res);
        Console.WriteLine("In reality, the minimum number needed is 3, i.e., 20 + 20 + 20");

        // Greedy: cannot ensure finding a global optimal solution
        coins = [1, 49, 50];
        amt = 98;
        res = CoinChangeGreedy(coins, amt);
        Console.WriteLine("\ncoins = " + coins.PrintList() + ", amt = " + amt);
        Console.WriteLine("The minimum number of coins required to make up " + amt + " is" + res);
        Console.WriteLine("In reality, the minimum number needed is 2, i.e., 49 + 49");
    }
}